### PR TITLE
Improve functionality of identity provider token retrieval

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
This PR intoduces a dedicated HTTP client for Keycloak operations, which is more accurately covers some cases like:
- Handling error messages from Keycloak and preserving original HTTP status codes;
- Fixes constant OAuth error when current user haven't  identity provider link

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/8288

#### Release Notes
Fix constant OAuth error when current user haven't  identity provider link 

#### Docs PR
N/A

### CODE REVIEW

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+import org.eclipse.che.dto.server.DtoFactory;
+import org.eclipse.che.multiuser.keycloak.shared.dto.KeycloakErrorResponse;
+import org.eclipse.che.multiuser.keycloak.shared.dto.KeycloakTokenResponse;
+
+/**
+ * Helps to perform keycloak operations and provide correct errors handling.
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class KeycloakServiceClient {
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java]: Should it be Singleton?

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+  private KeycloakSettings keycloakSettings;
+
+  private final Pattern assotiateUserPattern =
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java]: Shouldn't it be static?

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        .toString();
+  }
+
+
+  /**
+   * Gets auth token from given identity provider.
+   *
+   * @param oauthProvider provider name
+   * @return KeycloakTokenResponse token response
+   * @throws ForbiddenException
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java]: Looks like poor javadoc :smile_cat:

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+   */
+  private static String toJson(String source) {
+    Map<String, String> queryPairs = new HashMap<>();
+    Arrays.stream(source.split("&"))
+        .forEach(
+            p -> {
+              int delimiterIndex = p.indexOf("=");
+              queryPairs.put(p.substring(0, delimiterIndex), p.substring(delimiterIndex + 1));
+            });
+    return new Gson().toJson(queryPairs);
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java]: It think it'd better to have singleton GSON instance in static field

File: multiuser/keycloak/che-multiuser-keycloak-shared/src/main/java/org/eclipse/che/multiuser/keycloak/shared/dto/KeycloakTokenResponse.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ */
+package org.eclipse.che.multiuser.keycloak.shared.dto;
+
+import org.eclipse.che.dto.shared.DTO;
+
+/** @author Max Shaposhnik (mshaposh@redhat.com) */
+@DTO
+public interface KeycloakTokenResponse {
+
+  String getAccess_token();
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-shared/src/main/java/org/eclipse/che/multiuser/keycloak/shared/dto/KeycloakTokenResponse.java]: I think it's better to rewrite it in the following way:
```
@JsonFieldName("access_token")
String getAccessToken();
```

Comment (mshaposhnik) [reply multiuser/keycloak/che-multiuser-keycloak-shared/src/main/java/org/eclipse/che/multiuser/keycloak/shared/dto/KeycloakTokenResponse.java]: ups, forgot this trick, tnx

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+      String response = doRequest(url, HttpMethod.GET, null);
+      // Successful answer is not a json, but key=value&foo=bar format pairs
+      return DtoFactory.getInstance()
+          .createDtoFromJson(toJson(response), KeycloakTokenResponse.class);
+    } catch (BadRequestException e) {
+      if (assotiateUserPattern.matcher(e.getMessage()).matches()) {
+        // If user has no link with identity provider yet,
+        // we should threat this as unauthorized and send to oAuth login page.
+        throw new UnauthorizedException(e.getMessage());
+      } else {
```
Comment (akorneta) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java]: you can remove `else` here

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+            String.format(
+                "Failed access: %s, method: %s, response code: %d, message: %s",
+                UriBuilder.fromUri(url).replaceQuery("token").build(), method, responseCode, str));
+      }
+      final String contentType = conn.getContentType();
+      if (responseCode != HttpURLConnection.HTTP_NO_CONTENT
+          && contentType != null
+          && !(contentType.startsWith(MediaType.APPLICATION_JSON)
+              || contentType.startsWith("application/vnd.api+json"))) {
+        throw new IOException(conn.getResponseMessage());
```
Comment (akorneta) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java]: maybe it is better to throw `BadRequestException` with message like 'unsupported format' or add meaningful message, wdyt?

**Review by akorneta — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Comment by mshaposhnik** [comment]
Fixed;

**Pull request merged**